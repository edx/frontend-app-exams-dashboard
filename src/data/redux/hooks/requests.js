import { useSelector, useDispatch } from 'react-redux';

import * as redux from 'data/redux';
import * as module from './requests'; // eslint-disable-line import/no-self-import

const selectors = redux.selectors.requests;
const actions = redux.actions.requests;

export const statusSelector = selector => (requestName) => useSelector(selector(requestName));
export const useRequestIsPending = module.statusSelector(selectors.isPending);
export const useRequestIsFailed = module.statusSelector(selectors.isFailed);
export const useRequestIsCompleted = module.statusSelector(selectors.isCompleted);
export const useRequestIsInactive = module.statusSelector(selectors.isInactive);
export const useRequestError = module.statusSelector(selectors.error);
export const useRequestErrorCode = module.statusSelector(selectors.errorCode);
export const useRequestErrorStatus = module.statusSelector(selectors.errorStatus);
export const useRequestData = module.statusSelector(selectors.data);

export const useMakeNetworkRequest = () => {
  const dispatch = useDispatch();
  return ({
    requestKey,
    promise,
    onSuccess,
    onFailure,
  }) => {
    dispatch(actions.startRequest({ requestKey }));
    return promise.then((response) => {
      if (onSuccess) { onSuccess(response); }
      dispatch(actions.completeRequest({ requestKey, response }));
    }).catch((error) => {
      // Parse backend error from axios object
      let errorParsed = error;
      if (error?.customAttributes?.httpErrorResponseData) {
        errorParsed = JSON.parse(error.customAttributes.httpErrorResponseData);
      }
      if (onFailure) { onFailure(errorParsed); }
      dispatch(actions.failRequest({ requestKey, error: errorParsed }));
    });
  };
};

export const useClearRequest = (requestKey) => {
  const dispatch = useDispatch();
  return () => {
    dispatch(actions.clearRequest({ requestKey }));
  };
};
