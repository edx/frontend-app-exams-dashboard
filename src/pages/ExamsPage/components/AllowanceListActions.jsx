import { PropTypes } from 'prop-types';
import {
  ActionRow,
  Button,
  Icon,
  IconButtonWithTooltip,
  ModalDialog,
  useToggle,
} from '@openedx/paragon';
import { DeleteOutline, EditOutline } from '@openedx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import messages from '../messages';
import { useDeleteAllowance } from '../hooks';

const DeleteModal = (isOpen, onCancel, onDelete, formatMessage) => (
  <ModalDialog
    title="delete allowance"
    isOpen={isOpen}
    onClose={onCancel}
    variant="default"
    hasCloseButton
    isFullscreenOnMobile
  >
    <ModalDialog.Header>
      <ModalDialog.Title>
        {formatMessage(messages.deleteAllowanceHeader)}
      </ModalDialog.Title>
    </ModalDialog.Header>
    <ModalDialog.Body>
      <p>{formatMessage(messages.deleteAllowanceBody)}</p>
    </ModalDialog.Body>
    <ModalDialog.Footer>
      <ActionRow>
        <ModalDialog.CloseButton variant="tertiary" onClick={onCancel}>
          {formatMessage(messages.deleteAllowanceCancel)}
        </ModalDialog.CloseButton>
        <Button variant="primary" onClick={onDelete}>
          {formatMessage(messages.deleteAllowanceDelete)}
        </Button>
      </ActionRow>
    </ModalDialog.Footer>
  </ModalDialog>
);

const AllowanceListActions = ({ allowance }) => {
  const { formatMessage } = useIntl();
  const [isDeleteModalOpen, setDeleteModalOpen, setDeleteModalClosed] = useToggle(false);

  const deleteAllowance = useDeleteAllowance();

  const handleDelete = () => {
    deleteAllowance(allowance.id, setDeleteModalClosed);
  };

  return (
    <div className="allowances-actions text-right">
      <IconButtonWithTooltip
        tooltipPlacement="top"
        tooltipContent={formatMessage(messages.editAllowanceButton)}
        src={EditOutline}
        iconAs={Icon}
        alt={formatMessage(messages.editAllowanceButton)}
        onClick={setDeleteModalOpen}
        variant="primary"
        className="mr-2"
        size="sm"
      />
      <IconButtonWithTooltip
        tooltipPlacement="top"
        tooltipContent={formatMessage(messages.deleteAllowanceButton)}
        alt={formatMessage(messages.deleteAllowanceButton)}
        src={DeleteOutline}
        iconAs={Icon}
        onClick={setDeleteModalOpen}
        variant="secondary"
        size="sm"
        data-testid="delete-allowance-icon"
      />
      {DeleteModal(isDeleteModalOpen, setDeleteModalClosed, handleDelete, formatMessage)}
    </div>
  );
};

AllowanceListActions.propTypes = {
  allowance: PropTypes.shape({
    id: PropTypes.number.isRequired,
  }).isRequired,
};

export default AllowanceListActions;
