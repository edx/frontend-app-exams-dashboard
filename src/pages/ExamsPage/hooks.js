import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { RequestKeys } from 'data/constants';

import * as reduxHooks from 'data/redux/hooks';

import * as api from './data/api';
import * as selectors from './data/selectors';
import * as reducer from './data/reducer';

import * as module from './hooks'; // eslint-disable-line import/no-self-import

export const state = {
  // example of component state that does not need to be in redux
  exampleValue: (val) => React.useState(val), // eslint-disable-line
};

export const useFetchCourseExams = () => {
  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
  const dispatch = useDispatch();
  return (courseId) => (
    makeNetworkRequest({
      requestKey: RequestKeys.fetchCourseExams,
      promise: api.getCourseExams(courseId),
      onSuccess: (response) => dispatch(reducer.loadExams(response)),
    })
  );
};

export const useFetchExamAttempts = () => {
  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
  const dispatch = useDispatch();
  return (examId) => (
    makeNetworkRequest({
      requestKey: RequestKeys.fetchExamAttempts,
      promise: api.getExamAttempts(examId),
      onSuccess: (response) => dispatch(reducer.loadExamAttempts(response)),
    })
  );
};

export const useInitializeExamsPage = (courseId) => {
  const fetchCourseExams = module.useFetchCourseExams();
  React.useEffect(() => { fetchCourseExams(courseId); }, []); // eslint-disable-line react-hooks/exhaustive-deps
};

export const useExamsData = () => {
  const [exampleValue, setExampleValue] = state.exampleValue(0);
  const examsList = useSelector(selectors.courseExamsList);
  const currentExam = useSelector(selectors.currentExam);
  const isLoading = reduxHooks.useRequestIsPending(RequestKeys.fetchCourseExams);

  return {
    currentExam,
    examsList,
    isLoading,
    exampleValue,
    setExampleValue,
  };
};

export const useExamAttemptsData = () => {
  const attemptsList = useSelector(selectors.courseExamAttemptsList);
  return { attemptsList };
};
