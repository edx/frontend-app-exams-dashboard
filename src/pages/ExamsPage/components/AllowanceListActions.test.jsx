import { render, screen } from '@testing-library/react';

import { useDeleteAllowance } from '../hooks';
import AllowanceListActions from './AllowanceListActions';

jest.mock('../hooks', () => ({
  useDeleteAllowance: jest.fn(),
}));

// nomally mocked for unit tests but required for rendering/snapshots
jest.unmock('react');

const mockAllowance = {
  id: 1,
  examId: 1,
  examName: 'edX Exams',
  allowanceType: 'Minutes',
  extraTimeMins: 45,
};

describe('AllowanceListActions', () => {
  describe('snapshots', () => {
    test('displayed buttons should match snapshot', () => {
      expect(render(<AllowanceListActions allowance={mockAllowance} />)).toMatchSnapshot();
    });
  });

  it('should open the delete modal when the delete icon is clicked', () => {
    render(<AllowanceListActions allowance={mockAllowance} />);
    screen.getByTestId('delete-allowance-icon').click();
    expect(screen.getByText('Are you sure you want to delete this allowance?')).toBeInTheDocument();
  });

  describe('delete modal', () => {
    it('should delete the allowance when the delete button is clicked', () => {
      const mockDeleteAllowance = jest.fn();
      useDeleteAllowance.mockReturnValue(mockDeleteAllowance);
      render(<AllowanceListActions allowance={mockAllowance} />);
      screen.getByTestId('delete-allowance-icon').click();
      screen.getByText('Delete').click();
      expect(mockDeleteAllowance).toHaveBeenCalledWith(mockAllowance.id, expect.any(Function));
    });

    it('should close the modal when the cancel button is clicked', () => {
      render(<AllowanceListActions allowance={mockAllowance} />);
      screen.getByTestId('delete-allowance-icon').click();
      screen.getByText('Cancel').click();
      expect(screen.queryByText('Are you sure you want to delete this allowance?')).not.toBeInTheDocument();
    });

    it('should close the modal when the delete is successful', () => {
      const mockDeleteAllowance = jest.fn();
      useDeleteAllowance.mockReturnValue(mockDeleteAllowance);
      render(<AllowanceListActions allowance={mockAllowance} />);
      screen.getByTestId('delete-allowance-icon').click();
      screen.getByText('Delete').click();
      expect(screen.getByText('Are you sure you want to delete this allowance?')).toBeInTheDocument();
      mockDeleteAllowance.mock.calls[0][1]();
      expect(screen.queryByText('Are you sure you want to delete this allowance?')).not.toBeInTheDocument();
    });
  });
});
