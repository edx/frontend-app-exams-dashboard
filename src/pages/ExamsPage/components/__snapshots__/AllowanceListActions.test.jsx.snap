// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllowanceListActions snapshots displayed buttons should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="allowances-actions text-right"
      >
        <button
          aria-label="Edit allowance"
          class="btn-icon btn-icon-primary btn-icon-sm mr-2"
          type="button"
        >
          <span
            class="btn-icon__icon-container"
          >
            <span
              class="pgn__icon btn-icon__icon"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m14.06 9.02.92.92L5.92 19H5v-.92l9.06-9.06ZM17.66 3c-.25 0-.51.1-.7.29l-1.83 1.83 3.75 3.75 1.83-1.83a.996.996 0 0 0 0-1.41l-2.34-2.34c-.2-.2-.45-.29-.71-.29Zm-3.6 3.19L3 17.25V21h3.75L17.81 9.94l-3.75-3.75Z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
        </button>
        <button
          aria-label="Delete allowance"
          class="btn-icon btn-icon-secondary btn-icon-sm"
          data-testid="delete-allowance-icon"
          type="button"
        >
          <span
            class="btn-icon__icon-container"
          >
            <span
              class="pgn__icon btn-icon__icon"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 21h12V7H6v14zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4h-3.5z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="allowances-actions text-right"
    >
      <button
        aria-label="Edit allowance"
        class="btn-icon btn-icon-primary btn-icon-sm mr-2"
        type="button"
      >
        <span
          class="btn-icon__icon-container"
        >
          <span
            class="pgn__icon btn-icon__icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m14.06 9.02.92.92L5.92 19H5v-.92l9.06-9.06ZM17.66 3c-.25 0-.51.1-.7.29l-1.83 1.83 3.75 3.75 1.83-1.83a.996.996 0 0 0 0-1.41l-2.34-2.34c-.2-.2-.45-.29-.71-.29Zm-3.6 3.19L3 17.25V21h3.75L17.81 9.94l-3.75-3.75Z"
                fill="currentColor"
              />
            </svg>
          </span>
        </span>
      </button>
      <button
        aria-label="Delete allowance"
        class="btn-icon btn-icon-secondary btn-icon-sm"
        data-testid="delete-allowance-icon"
        type="button"
      >
        <span
          class="btn-icon__icon-container"
        >
          <span
            class="pgn__icon btn-icon__icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 21h12V7H6v14zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4h-3.5z"
                fill="currentColor"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
